### ISSUE TITLE
[ENH] Add a way to disable internal stacktraces

### ISSUE THREAD

**Issue by Chronial** [cython/cython#4450]
**Is your feature request related to a problem? Please describe.**
We use cython to wrap a C library. One of the things this wrapping layer does is turn error codes into exceptions. The traceback for that looks rather ugly:

```
  File "C:\dev\test.py", line 892, in run
    network.this_fails()
  File "library_cython.pyx", line 58209, in visum._visum.Network.this_fails
    _check(c.Visum_Network_this_fails(
  File "library_cython.pyx", line 26, in visum._visum._check
    raise visum.MyExceptionType(message)
visum.MyExceptionType(Stuff went wrong).
```

We would rather have:
```
  File "C:\dev\test.py", line 892, in run
    network.this_fails()
visum.MyExceptionType(Stuff went wrong).
```

**Describe the solution you'd like**
My understanding is that cython actively does some work to get these stack traces. Per default, the stacktrace would end at the last pure python frame, i.e. our desired result.
It would be great if cython had an option to disable its fancy stacktrace handling. Having this at the file level would be enough.

**Describe alternatives you've considered**
The only way I currently see to get our results is to not `raise` exceptions within cython, but to use the Python C Api Exception functions directly. But that would also only work if we did it in the outermost function of the API, because cython does the rewriting on function exit? That would complicate our code quite a bit and make it a lot harder to maintain.

**Comment by da-woods** [comment]
To me this seems like this is better done by using the existing Python exception control mechanisms rather than adding more Cython directives. E.g.

```
# in the outermost user-visible functions
try:
    g()
except BaseException as e:
    raise e.with_traceback(None)
```

You could write a decorator to generate this automatically.

**Comment by Chronial** [comment]
Unfortunately this workaround still leaves the `raise e.with_traceback` line in the traceback.

Also, this has unpleasant performance implications: Now we pay for the construction of traceback frames that are only going to be discarded. Using a decorator for this further adds overhead to every call, even if no exception is ever thrown. But using a decorator hand-written in C is the only possibility I see to actually get rid of the traceback.

Addressing this in cython seems a lot easier than the workarounds, since only a little bit of code needs not to be ran.

**Comment by Chronial** [comment]
Would you accept a PR for this that makes this a compile-time flag via a define or are you completely against supporting this in cython?

**Comment by da-woods** [comment]
> Would you accept a PR for this that makes this a compile-time flag via a define or are you completely against supporting this in cython?

I don't have the final say on this - that's ultimately up to @scoder.

The main argument against new options (either directives or compile-time flags) is that they add new combinations of things that can go wrong and thus new combinations that really need testing. Therefore the option ideally needs to adds something that can't be achieved in another way.

I slightly prefer a compile-time flag to a directive, but I'm not quite sure why.

Two other suggestions that might be worth knowing about:
* if Cython can't find the .pyx source file at runtime then it omits the details from the source-code.
* You might be able to get away with just `#define`ing `__Pyx_AddTraceback` to be empty? That wouldn't even require a PR (if it worked)

**Comment by Chronial** [comment]
> You might be able to get away with just #defineing __Pyx_AddTraceback to be empty? That wouldn't even require a PR (if it worked)

Nice idea, but unfortunately that doesn't work since it turns the function definition for `__Pyx_AddTraceback` into invalid syntax. I tried beeing clever with 
```
cdef extern from *:
    """#define __Pyx_AddTraceback(A, B, C, D)"""
```
at the beginning and 
```
cdef extern from *:
    """#undef __Pyx_AddTraceback"""
```
at the end of my cython file. But cython was even more clever and moved the blocks together to the top of the c output, making this a no-op.

But I managed to get results by doing a `re.sub(r'^\s*__Pyx_AddTraceback\(.*\);$', '', contents, flags=re.MULTILINE)` on the `cythonize` output. Thanks for your tip 👍 

But that is clearly an ugly hack :/. My understanding is that wrapping C APIs is a major use-case for cython and wrapped C APIs usually won't have any useful internal stack traces, So this seems like generally useful feature for cython to have.
